openedx/core/djangoapps/content/learning_sequences/api/tests/test_outlines.py

```diff
@@ -448,16 +468,123 @@ def setUpTestData(cls):
+
+    def enroll_student_in_course(self, student, course_key):
+        student.courseenrollment_set.create(course_id=course_key, is_active=True, mode="audit")
+
+    @override_waffle_flag(COURSE_ENABLE_UNENROLLED_ACCESS_FLAG, active=True)
+    def test_simple_outline(self):
+        """This outline is the same for everyone."""
+        STUDENTS = [
+            self.global_staff, self.student, self.student_but_not_enrolled_in_the_course, self.anonymous_user
+        ]
+        student_visibility_pairs = [
+            (student, visibility) for student in STUDENTS for visibility in CourseVisibility
+        ]
+
+        at_time = datetime(2020, 5, 21, tzinfo=timezone.utc)
+        for pair_i, pair in enumerate(student_visibility_pairs):
+            with self.subTest(pair=pair):
+                student, visibility = pair
+                course_key, _ = self.create_course_outline(visibility, pair_i)
+
+                if student == self.student:
+                    self.enroll_student_in_course(student, course_key)
+
+                user_course_outline = get_user_course_outline(course_key, student, at_time)
+                if student in [self.anonymous_user, self.student_but_not_enrolled_in_the_course]:
+                    # Both unenrolled student and anonymous student cannot access course outline when the course
+                    # visibility settings is set to private
+                    if visibility == CourseVisibility.PRIVATE:
+                        self.assertEqual(len(user_course_outline.sections), 0)
+                        self.assertEqual(len(user_course_outline.sequences), 0)
+                    else:
+                        # visibility is either public or public_outline
+                        self.assertEqual(len(user_course_outline.sections), 3)
+                        self.assertEqual(len(user_course_outline.sequences), 6)
+                        if visibility == CourseVisibility.PUBLIC_OUTLINE:
+                            self.assertTrue(
+                                all([
+                                    seq.usage_key not in user_course_outline.accessible_sequences
+                                    for seq in user_course_outline.sequences.values()
+                                ]),
+                                "Sequences shouldn't be accessible to anonymous or non-enrolled students "
+                                "for a public_outline_course"
+                            )
+                        else:
+                            # visibility is public
+                            self.assertTrue(
+                                all([
+                                    seq.usage_key in user_course_outline.accessible_sequences
+                                    for seq in user_course_outline.sequences.values()
+                                ]),
+                                "Sequences should be accessible to all users for a public course"
+                            )
```

Comment:
So while I gave that snippet above, I think it's clearer and simpler if you make three separate test methods with the different possible values of CourseVisibility, rather than having nested conditionals. The question that someone's going to come along and ask at some point is probably, "How is this supposed to work when course visibility is set to PUBLIC_OUTLINE?". It's a lot easier to see that if there's a test method that reads `test_public_outline` rather than having to trace these conditionals. It's okay to repeat some code if it makes the tests easier to understand (less nested, fewer abstractions).

Reply:
Good Tip :+1: